### DISCUSSION TITLE
How to reset sub-app data after each time step of main app?

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#18549] [Q&A General] [answered]
Dear all,
I am currently using multi-app feature in MOOSE to construct a multi-scale model that simulates the structural response of a component with reference to underlying micro-structural behaviour. The sub-app essentially contains an RVE modelled with strain periodicity approach. The workflow of this set-up requires that the stresses from region of interest in the main model is passed to the RVE, constraining the periodic boundaries and the resulting displacement (fluctuating) field is passed on to the main model enabling the determination of overall mechanical response. 

I am able to visualise this particular workflow under a multi-apps framework, transferring the required data back and forth between the models. While tyring out an example model, i observed that the sub-model seems to resume the analysis from the state attained at previous time step of main model, negating my requirement of a fresh start after each time step. I see that a sub-app reset can be performed at a specified time step of main app with the 'reset_apps' option (i guess that work only with 'TransientMultiApp'). Could any body advise me if this can be done at the end of each time step, so that the sub-app execution starts with a clean slate that ensures correct set of results? Attached here are the input files used and the log of execution.
Kind regards,
Arun
[Execution_log.txt](https://github.com/idaholab/moose/files/6946729/Execution_log.txt)
[macro.txt](https://github.com/idaholab/moose/files/6946730/macro.txt)
[sub_gs_uniaxial.txt](https://github.com/idaholab/moose/files/6946731/sub_gs_uniaxial.txt)

**Comment by GiudGiud** [comment]
Hello

Seems like what you want should be doable. While waiting for someone else to pitch in, could you please try:
- adding an initial condition in your subapp, see if it is obeyed at the next solve
- use the `no_backup_and_restore = true` parameter of the FullSolveMultiapp. It s meant for fixed point iterations, but those systems are intertwined a bit
- use the reset_apps and reset_time parameters. This is not just for TransientMultiapp, the times specified are the time steps for the main application. This should work but it's a little extreme. 

Finally, do you really need a FullSolveMultiApp with **a single time step** and a reset? It seems like this should either be a steady state solve in a FullSolveMultiapp OR a transient solve in a TransientMultiApp. Just an impression though

Guillaume

**Comment by abarun22** [comment]
Hi  Guillaume,
The inclusion of  an initial condition and the option 'no_backup_and_restore
= true' does not seem to affect the response. In principle I would
essentially want to solve a transient problem and so the usage of
'TransientMultiApp' would be more optimal for the computation. Inclusion of
'reset_apps' and 'reset_time' parameters in 'TransientMultiApp' does not
reset the sub-app data after every time step of the main model.
I've got my custom transfers (from and to the main model) to be performed
before and after execution of the sub-model.  The quantities that are
supposed to be transferred across are the stresses and displacements, which
are computed separately from the main and sub model respectively. Going by
these, I do not want any of the default transfers that are happening in
this workflow.

I really wonder what would be the best way to start the sub-app computation
in a clean state when the custom transfers are not implemented.
Kind regards,
Arun

**Reply by GiudGiud** [reply]
Thanks for trying these out. 
It doesnt reset even if you specify every time step in the list of time steps?

**Reply by abarun22** [reply]
I'm afraid we may not be able to provide a list of time steps to 'reset_time' option. And another problem i am facing now is that i could not re-produce the attainment of sub-app reset after the specified main app time step through this option. May be i am missing some thing. But nevertheless the objective now is that the reset is required to be functional on all time steps.

**Reply by GiudGiud** [reply]
You mean you can only provide a single value?

This would be a fairly easy modification to the framework to allow an option for all timesteps, or to be able to specify a list of timesteps. 
We would assist you in merging this capability in the framework if you developed it. If you are affiliated with a US university or national laboratory or other DOE subcontractors we could also just develop that for you.

**Reply by abarun22** [reply]
Yes, at present it can only take a single time value. I did tweak the code a little bit to allow for the reset to be happening for all the timesteps. Pasted here is the modified code block for all 'target_time' exceeding the specified reset time, without any reference to the reset history. 
  // First, see if any Apps need to be Reset
//  if (!_reset_happened && target_time + 1e-14 >= _reset_time)
    if (target_time + 1e-14 >= _reset_time)
  {
    _reset_happened = true;
    for (auto & app : _reset_apps)
      resetApp(app);
  }
This works ok for me at present. Now the question to be answered is whether you want the single reset or the bulk reset to be retained in the code. As you said the list (of reset_time) would be much better option to implement at this stage, given that it is more generic. I guess the modification in 'MultiApp.C' would be seen in all the related derived classes. I am happy to do the modification if you are ok with it.
Kind regards,
Arun

**Reply by GiudGiud** [reply]
We would be very interested in supporting the list of reset_time.
The single reset time was likely developed to support another application, so we cannot switch the behavior from single to all like this though.

**Comment by abarun22** [comment]
Hi Guillaume,
Here is the modified code block for enabling the variable '_reset_time' as a list of times from the main app. 
void
MultiApp::preTransfer(Real /*dt*/, Real target_time)
{
  if ((_reset_time.size() == 0))
    mooseError("Please specify the list comprising times.");

   // First, see if any Apps need to be Reset
   for (unsigned int i=0; i<_reset_time.size();i++){
     if (target_time == _reset_time[i])
  {
    for (auto & app : _reset_apps)
      resetApp(app);
  }
}
I've tested that only with TransientMultiApp, while other objects are yet to be tested. Please let me know if this can be approved through a PR?
Kind regards,
Arun

**Reply by GiudGiud** [reply]
I think this 
``
if ((_reset_time.size() == 0))
mooseError("Please specify the list comprising times.");
``
makes reset times mandatorym which we dont necessarily want. 

Is there any logic with reset_times than makes sure that we hit all of them exactly? Or if the time is not reached exactly then the reset is skipped? What did you see in your testing?

**Reply by abarun22** [reply]
Yes the chosen logic will compare each element in the list to the current time 'target_time' making sure that all of those in `reset_time` are addressed. In the other case, where the specified times are not reached during the course of computation (for example due to time step cut by auto time stepping scheme) we may not be able to reset the sub-app following that time step. These things are subjective and cannot be known at the time of preparing the input commands. I welcome your suggestions in that regard. Attached here are the example input files i used earlier for testing this feature. 
[18082021.zip](https://github.com/idaholab/moose/files/7008386/18082021.zip)

**Reply by GiudGiud** [reply]
Ok my question was more the second point. This is really not good but it's also not because of your change, it seems to be an existing problem with the reset_time parameter. 
For comparing the target time and the reset times, you cannot use ==, this will be vulnerable to floating point errors. Use the 
 absoluteFuzzyEqual() routine.
Otherwise, please make the PR at your convenience

**Reply by GiudGiud** [reply]
ok actually the existing code behaves differently
```
  // First, see if any Apps need to be Reset
  if (!_reset_happened && target_time + 1e-14 >= _reset_time)
  {
    _reset_happened = true;
    for (auto & app : _reset_apps)
      resetApp(app);
  }
  ```
  you need to mimick that behavior (reset after time has passed) rather than try to hit the times exactly

**Comment by friedmud** [comment, marked as answer]
Before going futher here... I would like to know more about your application.  Are you doing a steady state solve in the sub-app?  If so, it shouldn't matter what solution was there before.  Are you trying to do a transient run in the sub-app at every timestep?

Some more information would be useful.

**Reply by friedmud** [reply]
@abarun22 I really don't think you need to be doing anything with resetting apps.  That would be a _very_ rare use-case.  Can you please explain the physical system you're trying to solve?

**Reply by abarun22** [reply]
Dear Derek/Guillaume,
Thanks for your comments. Attached please find a presentation that describes the physics i am trying to solve. In my opinion, i would prefer a sub-app reset as that would allow me to start the sub-model computation from a clean state, thus enabling the custom transfers i am planning to implement, taking shape. The sub-app essentially mimics a steady state analysis though. 
[CPFEM_descr_19Aug2021.pptx](https://github.com/idaholab/moose/files/7013570/CPFEM_descr_19Aug2021.pptx)

Regarding the `reset_time` modification, i still feel that hitting the right time would be the correct approach, as the current approach uses the precision value in comparing the times, which in our case is by-passed by the usage of absoluteFuzzyEqual(). Here is that code.

<pre><code> void
MultiApp::preTransfer(Real /*dt*/, Real target_time)
{
  if ((_reset_time.size() == 0))
    mooseError("Please specify the list comprising times.");

   // First, see if any Apps need to be Reset
//  if (!_reset_happened && target_time + 1e-14 >= _reset_time)
   for (unsigned int i=0; i<_reset_time.size();i++){
     if (MooseUtils::absoluteFuzzyEqual(target_time, _reset_time[i]))
  {
//    _reset_happened = true;
    for (auto & app : _reset_apps)
      resetApp(app);
  }
}
</code></pre>

Otherwise this code block does the right job, still resetting the required times.
  
Kind regards,
Arun

**Reply by GiudGiud** [reply]
When using a FullSolveMultiapp, you usually do NOT want to start from a clean state, because the result is usually the same whether you start from a clean state or from a fresh case, it just takes longer to get to the same place.

In your current configuration of FullSolveMultiApp + Transient with 1 timestep, you are getting a different result if you re-initialize, which was why I was helping you with the reset. 

If your subapp is mimicking a steady state analysis, why are you not using a Steady executioner to try to get to steady state?
For a steady subapp, definitely no need to reset. 

This code block does not reset if you go over the reset_time (without hitting it exactly or even having a timestep near it). So it doesnt work for what people have used it for and I think it may fail the test suite.

**Reply by abarun22** [reply]
Yes i think `FullSolveMultiapp` runs the sub-model independently from the main model and so we will not be required to reset the sub-app as we did for the `TransientMultiApp`. However `FullSolveMultiapp` produces a segmentation violation (at the end of second time step of main app), even with the original case of a single value `reset_time` and so i could not check the state of sub-app computation in the subsequent time steps.  From my perspective i will intend to use `FullSolveMultiapp` given the computation runs OK without seg faults. As it stands, i shall continue using `TransientMultiApp` with Transient runs in both apps and that gives acceptable results with manual imposition of list of reset times. Now that i know the sub-app starts off in a clean manner, i shall then think about writing my own transfer functions to push and pull the data between the apps.

As per your suggestion i've also modified the code further to perform a reset even at odd times decided by the time stepping scheme. Here is the new block of code.
<pre><code>
   for (unsigned int i=0; i<_reset_time.size();i++){
      if (target_time + 1e-14 >= _reset_time[i])
  {
    for (auto & app : _reset_apps)
      resetApp(app);
      break;
  }
</code></pre>

Also find attached the new set of input files relevant to my current work, which will form the building block for further developments.
[20082021.zip](https://github.com/idaholab/moose/files/7022517/20082021.zip)

Kind regards,
Arun

**Reply by GiudGiud** [reply]
Hi Arun

This new piece of code will continuously reset the apps. You need to keep track of when each reset happened, using a vector of boolean. The original code there should guide you into what this system should be able to do.

A seg fault is very concerning. Could you please compile in debug mode, `METHOD=dbg make -j #cores` then get us a backtrace for this error, either by running with start-in-debugger, or with `gdb --ex run --ex bt --args executable-dbg -i input_file.i` 

Also there are a number of existing transfers. Does none of them work for your case? What are you trying to transfer?

**Reply by abarun22** [reply]
Hi Guillaume,
Here is the new segment of code that does it correctly. This will reset not only the times given in the list, but also the others which arise in-between them. Essentially we check if the `target_time` falls in between the consecutive `reset_time's` in the list and reset those target times. Please take a look at it and let me know if that's ok with you.

<pre><code>
   // Sort the 'reset_time' list in increasing order
   std::sort(_reset_time.begin(), _reset_time.end());

   // Reset apps for the times provided in the list
   if (_reset_time.size()>1){
      for (unsigned int i=0; i<_reset_time.size()-1;i++){
        if (target_time + 1e-14 >= _reset_time[i] && target_time < _reset_time[i+1]){
          for (auto & app : _reset_apps)
          resetApp(app);
        }
      }
  }

  // Reset apps for the following cases:
  // 1) list contains a single entry
  // 2) Final entry in the list which was not reset previously

  if ((_reset_time.size()==1 && target_time==_reset_time[0]) || target_time == _reset_time[_reset_time.size()-1]){
        for (auto & app : _reset_apps)
          resetApp(app);
  }
</code></pre>

I am yet to look in to the seg fault issues with `FullSolveMultiapp`. Will look in to that and let you know the status soon. Regarding the transfer, all of the available schemes seems to couple certain field variables. None of them allows me to transfer just the data that is required to be visible to the other app. I am interested to transfer the following: 
1) original/deformed coordinates of main model - From main app to sub app
2) Stress tensor of the heavily stresses element - From main app to sub app
3) Updated displacements of the sub-model to be linked with the equivalent displacements in main model - From sub-app to main app
I presume that the data from the first two cases be packaged in to a user object, which can in some way be transferred to the sub-app. The sub-app in turn may need a receiver class to fetch the data and make it ready for the sub-app space. The displacements can be coupled through a standard scheme such as 'MultiAppNearestNodeTransfer' if i am correct. Eager to know your suggestions in this regard.
Kind regards,
Arun

**Reply by GiudGiud** [reply]
Hi Arun

This still does not do the trick. You don't keep track of which reset has happened, so you ll keep resetting indefinitely.
We also do NOT need to separate the case with more than one reset from the case with a single reset time. 
You need to look at the original code and understand why that boolean, keeping track of when the sub app reset was performed, was there in the first place. 

Ok I'd like to see that backtrace for the seg fault. 

@dschwen is there any pre-made transfers/user objects for transfering stress tensors to a sub app? Or should they move the data to an aux-variable then use a regular transfer?
Are you going to use 3. to try to deform the main app mesh? Or to compute some quantity in the main app?

Guillaume

**Reply by abarun22** [reply]
Hi Guillaume,
I guess this time we may be ok. The code below might do the trick.

<pre><code>
   // Sort the 'reset_time' list in increasing order
   std::sort(_reset_time.begin(), _reset_time.end());

  // First, see if any Apps need to be Reset
     for (unsigned int i=0; i<_reset_time.size();i++){
        if (!_reset_happened[i] && (target_time + 1e-14 >= _reset_time[i]))
       {
        _reset_happened[i] = true;
        for (auto & app : _reset_apps)
           resetApp(app);
       }
     }
</code></pre>
I tried this before, but was overshadowed by the odd behaviour where the time step changes drastically for each length of list we provide in the input file. I wonder what is that causing the dependency of time step cut on the reset process. Strictly speaking the time step should not be influenced by the number of elements provided in the `reset_time` list. 
I will give you my seg fault results possibly in my next post. 
Yes, i calculate the updated displacements of the main app mesh inside the sub app and transfer that on to the main app with an intention to overwrite the old displacements, that will then feed on to the next time step.
Kind regards,
Arun

**Reply by GiudGiud** [reply]
Yup this looks good. We also would not want to reset the app multiple times if two reset_times have been hit
So let's have another, single, boolean to track that

Umm this might be difficult. `disp_x` is not an AuxVariable, it s a main variable right? In which case you can't modify it like this

**Reply by abarun22** [reply]
With this approach, the boolean array is enough to track the status of each reset_time in the list, whether the reset is accomplished or not. I do not see any reason why an app should be reset multiple times here. Could you please brief me a bit more here and help me understand the problem deeply?
Yes i am interested in overwriting disp_x which is the main variable. I do not know why this can't be modified.

**Reply by GiudGiud** [reply]
it can be reset multiple times with different reset times on the same timestep
Let s say time goes from 10 to 14 and there are resets at 11 and 12, then this loop will trigger them both. 

A main variable is the solution of the non linear problem. I dont think we allow modifying it directly, outside of initial conditions, but I may be wrong

**Reply by abarun22** [reply]
Lets consider a time constant time step of 1 for this example. A possibility of double trigger will occur only when the current (target_time) time reaches or exceeds a value of 12. When the target time reaches 12 the boolean array `_reset_happened[]` would've been set {1, 1} and the code will hit only the reset time 12. Also at this target time, reset time 11 will not be hit as the first element of boolean array was already set to `true` The subsequent times 13 and 14 would never hit the reset_values {11, 12} as both elements of boolean array are already set to `true`. The dependency of actual times on the length of reset time list reported earlier may result from some other reasons.

**Reply by GiudGiud** [reply]
please think about my scenario. 
Yours has the desired behavior but it s not what I m talking about

**Reply by abarun22** [reply]
Tried to run the debugger and received the following backtrace
<pre><code>
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff27d58a1 in FullSolveMultiApp::solveStep (this=0xe0e8e0, auto_advance=true)
    at /home/abalasub/Working/Projects/moose/framework/src/multiapps/FullSolveMultiApp.C:126
126         ex->execute();
#0  0x00007ffff27d58a1 in FullSolveMultiApp::solveStep (this=0xe0e8e0, auto_advance=true)
    at /home/abalasub/Working/Projects/moose/framework/src/multiapps/FullSolveMultiApp.C:126
#1  0x00007ffff1edb202 in FEProblemBase::execMultiApps (this=0xc6a200, type=..., auto_advance=true)
    at /home/abalasub/Working/Projects/moose/framework/src/problems/FEProblemBase.C:4303
#2  0x00007ffff2470122 in PicardSolve::solveStep (this=0xce7d10, begin_norm_old=1.7976931348623157e+308,
    begin_norm=@0x2fd51d0: 0, end_norm_old=1.7976931348623157e+308, end_norm=@0x2fc6ed0: 0, relax=false, relaxed_dofs=...)
    at /home/abalasub/Working/Projects/moose/framework/src/executioners/PicardSolve.C:490
#3  0x00007ffff246e6f1 in PicardSolve::solve (this=0xce7d10)
    at /home/abalasub/Working/Projects/moose/framework/src/executioners/PicardSolve.C:241
#4  0x00007ffff213b666 in TimeStepper::step (this=0x11dadc0)
Missing separate debuginfos, use: debuginfo-install bzip2-libs-1.0.6-13.el7.x86_64 elfutils-libelf-0.166-2.el7.x86_64 elfutils-libs-0.166-2.el7.x86_64 glibc-2.17-222.el7.x86_64 glpk-4.52.1-2.el7.x86_64 libX11-1.6.5-1.el7.x86_64 libXau-1.0.8-2.1.el7.x86_64 libattr-2.4.46-12.el7.x86_64 libcap-2.22-8.el7.x86_64 libxcb-1.12-1.el7.x86_64 systemd-libs-219-67.el7.x86_64 xz-libs-5.2.2-1.el7.x86_64
</pre><\code>

**Reply by abarun22** [reply]
May be i am missing some thing here. Lets go much deeper here. Could you let me know at what target time you might experience a double trigger in this example? In my opinion each target time will only hit one item in the list. In case if it sees other items, the boolean array restrict its entry in to the reset part of code.

**Reply by GiudGiud** [reply]
The timestepper wont stop at target time 11 and 12, it goes from 10 to 14. Reaching 14, it ll reset twice. 

You could also add a break statement, it does not have to be a boolean. 

This backtrace is very strange. It s as if the executioner was null. Is this a recent version of moose?

**Reply by abarun22** [reply]
Yes, that's all we need for the fix. Added a break statement as you suggested.
<pre><code>
   // Sort the 'reset_time' list in increasing order
   std::sort(_reset_time.begin(), _reset_time.end());

  // First, see if any Apps need to be Reset
     for (unsigned int i=0; i<_reset_time.size();i++){
        if (!_reset_happened[i] && (target_time + 1e-14 >= _reset_time[i]))
       {
        _reset_happened[i] = true;
        for (auto & app : _reset_apps)
           resetApp(app);
        break;
       }
     }
</pre></code>

Yes this was sort of recent version committed earlier this year. Here is the commit status if you like.

commit 735703d125b31643a9bf50201239091abf750c97
Merge: d749de5 b1a1f9a
Author: Ben Spencer <benjamin.spencer@inl.gov>
Date:   Fri Jan 8 20:02:46 2021 -0700
Kind regards,
Arun

**Reply by GiudGiud** [reply]
January is fairly old for debugging. There are 8 months of commits to look at for a possible fix. I would update.
